src/Hosting/Hosting/src/Internal/StartupLoader.cs

```diff
@@ -204,15 +205,13 @@ void RunPipeline(TContainerBuilder containerBuilder)
                 {
                     var filters = HostingServiceProvider
 #pragma warning disable CS0612 // Type or member is obsolete
-                        .GetRequiredService<IEnumerable<IStartupConfigureContainerFilter<TContainerBuilder>>>()
+                        .GetRequiredService<IEnumerable<IStartupConfigureContainerFilter<TContainerBuilder>>>();
 #pragma warning restore CS0612 // Type or member is obsolete
-                        .Reverse()
-                        .ToArray();
 
                     Action<TContainerBuilder> pipeline = InvokeConfigureContainer;
-                    for (int i = 0; i < filters.Length; i++)
+                    foreach (var filter in filters.Reverse())
```

Comment:
Why change this to use foreach + Reverse?

Reply:
After removing the invocation of `ToArray()`, it made more sense to me to move the invocation of `Reverse()` to its usage.

Reply:
I'm sorry, but I don't get your point.

Reply:
The goal is to avoid allocations in Reverse right? Why not remove it and loop through the collection backwards?

Reply:
Sorry! It was a typo due to copy+paste.

In this case the goal is to avoid `Enumerable.ToArray()`.

Reply:
Why are you worried about ToArray in one scenario and Reverse in the other? Shouldn't these use the same approach?

Reply:
It's not about one or the other. It's about both.

In one case, the size of the collection is checked, so I kept `ToArray` to check its size. Now I don't need to reverse it because I can iterate it in reverse order.

In the other case, the collection is only iterated, so `Reverse` serves the purpose.

Reply:
Now that you mention it, this is a pattern I often use:

```
var filters = HostingServiceProvider.GetRequiredService<IEnumerable<IStartupConfigureServicesFilter>>()
    .Reverse();
var enumerator = filters.GetEnumerator();
try
{
    // If there are no filters just run startup (makes IServiceProvider ConfigureServices(IServiceCollection services) work.
    if (!enumerator.Next())
    {
        return startup(services);
    }
    Action<IServiceCollection> pipeline = InvokeStartup;
    do
    {
        pipeline = enumerator.Current.ConfigureServices(pipeline);
    } while (enumerator.Next());
    pipeline(services);
    // We return null so that the host here builds the container (same result as void ConfigureServices(IServiceCollection services);
    return null;
}
finally
{
    
    enumerator.Dispose();
}
```

No `ToArray`!

Reply:
`filters.Length == 0` could be an `.Any()` check instead.

Reply:
NO! Don't do that!

If the type provides a `Length`/`Count`, use it. Otherwise, there will be interface implementation checks and an enumerator allocation.

Now, if you go with `Reverse` and use `Any` you'll pay that cost when you can avoid it with [this code](https://github.com/dotnet/aspnetcore/pull/28387#discussion_r538772791).